File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/directionprocessor/BlockingDirectionsProcessor.java
package co.electric.snake.strategy.bonostrategy.directionprocessor;

import co.electric.snake.framework.model.Arena;
import co.electric.snake.framework.model.Coordinate;
import co.electric.snake.framework.model.Direction;
import co.electric.snake.framework.model.Snake;
import co.electric.snake.strategy.bonostrategy.BlockingDirectionContainer;
import co.electric.snake.strategy.bonostrategy.SimpleDirectionContainer;
import co.electric.snake.strategy.bonostrategy.distanceprocessor.DistanceProcessorChain;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class BlockingDirectionsProcessor {

    private static final Logger LOG = LoggerFactory.getLogger(BlockingDirectionsProcessor.class);

    private final DistanceProcessorChain distanceProcessorChain;

    public BlockingDirectionsProcessor(DistanceProcessorChain distanceProcessorChain) {
        this.distanceProcessorChain = distanceProcessorChain;
    }

    public BlockingDirectionContainer getDirections(Snake snake, Arena arena, SimpleDirectionContainer filteredDirections) {
        final BlockingDirectionContainer blockingDirections = new BlockingDirectionContainer();
        final Coordinate actualHeadCoordinate = snake.getHeadCoordinate();
        final Coordinate maxCoordinate = arena.getMaxCoordinate();
        for (Direction actualDirection : filteredDirections) {
            Coordinate coordinateToInvestigate = actualHeadCoordinate;
            final int maxCoordinateForDirection = getMaxCoordinateForDirection(maxCoordinate, actualDirection);
            for (int i = 0; i < maxCoordinateForDirection; i++) {
                coordinateToInvestigate = arena.nextCoordinate(coordinateToInvestigate, actualDirection);
                if (arena.isOccupied(coordinateToInvestigate)) {
                    final Snake blockingSnake = getBlockingSnake(arena, coordinateToInvestigate);
                    final int blockingTailLength = getBlockingTailLength(blockingSnake, coordinateToInvestigate);
                    final int distanceToBlock = getDistanceToBlock(actualDirection, actualHeadCoordinate, coordinateToInvestigate, maxCoordinate);
                    if (isBlockingRisk(blockingTailLength, distanceToBlock)) {
                        blockingDirections.putData(actualDirection, coordinateToInvestigate, distanceToBlock);
                    }
                }
            }
        }

        LOG.info("Blocking Directions: " + blockingDirections);

        return blockingDirections;
    }

    private int getDistanceToBlock(Direction actualDirection, Coordinate actualHeadCoordinate, Coordinate coordinateToInvestigate, Coordinate maxCoordinate) {
        return distanceProcessorChain.getDistance(actualDirection, actualHeadCoordinate, coordinateToInvestigate, maxCoordinate);
    }

    private int getMaxCoordinateForDirection(Coordinate maxCoordinate, Direction actualDirection) {
        final int maxCoordinateForDirection;
        if (Arrays.asList(Direction.NORTH, Direction.SOUTH).contains(actualDirection)) {
            maxCoordinateForDirection = maxCoordinate.getX() - 1;
        } else {
            maxCoordinateForDirection = maxCoordinate.getY() - 1;
        }
        return maxCoordinateForDirection;
    }

    private Snake getBlockingSnake(Arena arena, Coordinate blockingCoordinate) {
        return arena.getSnakesInNewList().stream()
                .filter(snake -> snake.getBodyItemsInNewList().stream().anyMatch(bodyItem -> bodyItem.equals(blockingCoordinate)))
                .findAny()
                .orElse(null);
    }

    private int getBlockingTailLength(Snake blockingSnake, Coordinate blockingCoordinate) {
        final AtomicInteger blockingTailLength = new AtomicInteger();
        final AtomicBoolean reachedTheBlockingPart = new AtomicBoolean(false);
        blockingSnake.getBodyItemsInNewList().forEach(
                actualBodyItem -> {
                    if (actualBodyItem.equals(blockingCoordinate)) {
                        reachedTheBlockingPart.set(true);
                    }
                    if (reachedTheBlockingPart.get()) {
                        blockingTailLength.getAndIncrement();
                    }
                }
        );
        return blockingTailLength.get();
    }

    private boolean isBlockingRisk(int blockingTailLength, int distanceToBlock) {
        return blockingTailLength >= distanceToBlock;
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/directionprocessor/BlockingDirectionsProcessor.kt
package co.electric.snake.strategy.bonostrategy.directionprocessor

import co.electric.snake.framework.model.Arena
import co.electric.snake.framework.model.Coordinate
import co.electric.snake.framework.model.Direction
import co.electric.snake.framework.model.Snake
import co.electric.snake.strategy.bonostrategy.BlockingDirectionContainer
import co.electric.snake.strategy.bonostrategy.SimpleDirectionContainer
import co.electric.snake.strategy.bonostrategy.directionprocessor.distanceprocessor.DistanceProcessorChain
import org.slf4j.LoggerFactory

class BlockingDirectionsProcessor(private val distanceProcessorChain: DistanceProcessorChain) {

    companion object {
        private const val SAFE_THRESHOLD = 25

        private val LOG = LoggerFactory.getLogger(BlockingDirectionsProcessor::class.java)
    }

    fun getDirections(snake: Snake, arena: Arena, directionsToUse: SimpleDirectionContainer): BlockingDirectionContainer {
        val blockingDirections = BlockingDirectionContainer()
        val headCoordinate = snake.getHeadCoordinate()
        val maxCoordinate = arena.maxCoordinate
        directionsToUse.forEach {
            var coordinateToInvestigate = headCoordinate
            val maxCoordinateForDirection = getMaxCoordinateForDirection(maxCoordinate, it)
            for (i in 0 until maxCoordinateForDirection) {
                coordinateToInvestigate = arena.nextCoordinate(coordinateToInvestigate, it)
                if (arena.isOccupied(coordinateToInvestigate)) {
                    val blockingSnake = getBlockingSnake(arena, coordinateToInvestigate)
                    val blockingTailLength = getBlockingTailLength(blockingSnake, coordinateToInvestigate)
                    val distanceToBlock = getDistanceToBlock(it, headCoordinate, coordinateToInvestigate, maxCoordinate)
                    if (distanceToBlock <= SAFE_THRESHOLD && isBlockingRisk(blockingTailLength, distanceToBlock)) {
                        blockingDirections.putData(it, coordinateToInvestigate, distanceToBlock)
                    }
                }
            }
        }
        LOG.info("Blocking Directions: $blockingDirections")
        return blockingDirections
    }

    private fun getMaxCoordinateForDirection(maxCoordinate: Coordinate, direction: Direction): Int {
        return if (setOf(Direction.NORTH, Direction.SOUTH).contains(direction)) {
            maxCoordinate.x - 1
        } else {
            maxCoordinate.y - 1
        }
    }

    private fun getBlockingSnake(arena: Arena, blockingCoordinate: Coordinate): Snake {
        return arena.getSnakesInNewList().stream()
                .filter { snake -> snake.getBodyItemsInNewList().stream().anyMatch { it == blockingCoordinate } }
                .findAny()
                .orElse(null)
    }

    private fun getBlockingTailLength(blockingSnake: Snake, blockingCoordinate: Coordinate): Int {
        var blockingTailLength = 0
        var reachedTheBlockingPart = false
        blockingSnake.getBodyItemsInNewList().forEach {
            if (it == blockingCoordinate) {
                reachedTheBlockingPart = true
            }
            if (reachedTheBlockingPart) {
                blockingTailLength++
            }
        }
        return blockingTailLength
    }

    private fun getDistanceToBlock(direction: Direction, headCoordinate: Coordinate, blockingCoordinate: Coordinate, maxCoordinate: Coordinate): Int {
        return distanceProcessorChain.getDistance(direction, headCoordinate, blockingCoordinate, maxCoordinate)
    }

    private fun isBlockingRisk(blockingTailLength: Int, distanceToBlock: Int): Boolean {
        return blockingTailLength >= distanceToBlock
    }

}
